hail/src/main/scala/is/hail/io/vcf/LoadVCF.scala

```diff
@@ -1744,7 +1744,7 @@ object MatrixVCFReader {
 
     LoadVCF.warnDuplicates(sampleIDs)
 
-    new MatrixVCFReader(params, fileListEntries, referenceGenome, header1)
+    new MatrixVCFReader(params.copy(files = fileListEntries.map(_.getPath)), fileListEntries, referenceGenome, header1)
```

Comment:
This feels inconsistent across the readers.

`StringTableReader` is special-cased in `SemanticHash` to use the `fileStatuses` but its `params.files` is still the original glob-pattern-containing params.

`pathsUsed` as introduced [a long time ago](https://github.com/hail-is/hail/pull/8327) to protect against silly pipelines like:
```python3
hl.read_matrix_table('foo.mt').write('foo.mt')
```
This works fine for non-globbed data sources like matrix tables and tables. However, even in the original PR #8327, it seems we would always fail to notice:
```python3
hl.export_vcf(hl.import_vcf('chr*.vcf'), 'chr1.vcf')
```

It seems to me that the durable & reliable fix to this is for `pathsUsed` to always be the list of post-globbed files/blobs. I have a mild preference to treat `params` as an immutable record of what the user requested.

Taking this one step further: `pathsUsed` could be a list of `FileStatus` (which is a super type of `FileListEntry`) and `FileStatus` could have an `eTag`. This avoids O(N_FILES) calls to google to get the etag of all the files. Any reader that uses `glob` will already have a `FileListEntry` for every file. I suspect readers that don't use glob *do* still check the existence of the files ahead of time. They should just keep a `FileStatus` (which is proof of existence anyway) around so that we can later get its etag.